src/middlewared/middlewared/plugins/vm.py

```diff
@@ -515,6 +524,64 @@ def flags(self):
             self.logger.error("===> {0}".format(err))
             return False
 
+    @private
+    async def find_clone(self, name):
+        data = await self.middleware.call('vm.query', [], {'order_by': ['name']})
+        clone_index = 0
+        next_name = ""
+        for vm_name in data:
+            if name in vm_name['name'] and '_clone' in vm_name['name']:
+                name_index = int(vm_name['name'][-1])
+                next_name = vm_name['name'][:-1]
+                if name_index >= clone_index:
+                    clone_index = int(name_index) + 1
+
+        if next_name:
+            next_name = next_name + str(clone_index)
+        else:
+            next_name = name + '_clone' + str(clone_index)
+
+        return next_name
+
+    @accepts(Int('id'))
+    async def clone(self, id):
+        vm = await self._manager.clone(id)
+
+        if vm is None:
+            self.logger.error("Cannot clone a VM that does not exist.")
```

Comment:
The way I see it cloning an nonexistent VM should raise an error. How about raise CallError('VM does not exist', errno.ENOENT) ?
In my opinion make it more explicit, without having to look at logs.